src/city_score/sources/traillink.py

```diff
@@ -0,0 +1,87 @@
+from ..cache import cache
+from ..decorators import dimension, scorer
+from ..source import Source
+from html.parser import HTMLParser
+import requests
+
+activities_filter = {
+    'Atv': 'ATV',
+    'Bike': 'BIKE',
+    'Birding': 'BIRD',
+    'Cross Country Skiing': 'XSKI',
+    'Dog Walking': 'DOG',
+    'Fishing': 'FISH',
+    'Geocaching': 'GEO',
+    'Hiking': 'HIKE',
+    'Horseback Riding': 'HORSE',
+    'Inline Skating': 'SKTS',
+    'Mountain Biking': 'MTBK',
+    'Running': 'RUN',
+    'Snowmobiling': 'SNOW',
+    'Walking': 'WALK',
+    'Wheelchair Accessible': 'WHEEL'
+}
+
+class TrailLinkMilesParser(HTMLParser):
+    def __init__(self):
+        super().__init__()
+        self.found = False
+        self.total_trail_miles = 0
+
+    def handle_starttag(self, tag, attrs):
+        if tag == 'span':
+            for attr in attrs:
+                if attr[0] == 'itemprop' and attr[1] == 'distance':
+                    self.found = True
+
+    def handle_data(self, data):
+        if self.found:
+            self.total_trail_miles += float(data.split()[0])
+
+    def handle_endtag(self, tag):
+        if tag == 'span' and self.found:
+            self.found = False
+
+class TrailLink(Source):
+    base_url = 'https://www.traillink.com/trailsearch'
+
+def format(number):
+    return f'{number} miles'
+
+@dimension('TrailLink Mileage')
+def trail_total_miles(city, activities=[], min_length=5):
+    key = 'traillink-%s-%s' % (str(city), str("-".join(activity.replace(" ", "") for activity in activities)))
+    count = cache.get(key)
+    if count is not None:
+        return format(count)
+    
+    activity_values = []
+    for activity in activities:
+        activity_values.append(activities_filter[activity])
+
+    params={
+        'city': city.name,
+        'state': city.state,
+        'activities': ",".join(activity_values),
+        'length': f'{min_length}|99999'
+    }
+    response = requests.get(TrailLink.base_url, params)
+
+    if response.status_code == 200:
+        parser = TrailLinkMilesParser()
+        parser.feed(response.text)
+        count = int(parser.total_trail_miles)
+        cache.set(key, count)
+        return format(count)
+
+@scorer('TrailLink')
+def trail_total_miles_scorer(city, lower, upper, activities=[]):
+    data = trail_total_miles(city, activities)
+    if data == '?': return 0
+
+    count = float(data.split(' ')[0])
```

Comment:
To avoid these previous two lines, you could create one primary function that does the "work" and the dimension (formats the value) and scorer (returns `0` if `None`) invoke it and just do "their" work.

Reply:
updated https://github.com/jacobbudin/city-score/pull/7/commits/cd1ecb8acc6f0ed8ecacc86a62e1f5ab5360b4a1 - lmk if thats not at all what you were talking about 😅 . the dimension is the thing that formats the value to be displayed in the table

don't think i needed the `?` string here anyway (could make sense if the status_code != 200 i guess). adding any filters to the query will just return `0 miles` (if there are no trails for example)

Reply:
Yes, exactly. Two small comments:
1. It's `scrape`, not `scrap`. 🗑️
2. Unless I'm missing something, you'll want to put the caching inside your new function so the remaining functions both can use it.

Reply:
1. lol woops
2. 👍